src/cargo/ops/cargo_info/mod.rs

```diff
@@ -0,0 +1,398 @@
+                .map(|v| v.is_compatible_with(rustc_version))
+                .unwrap_or_else(|| false);
+            // MSRV compatible version is preferred.
+            match (s1_matches, s2_matches) {
+                (true, false) => std::cmp::Ordering::Greater,
+                (false, true) => std::cmp::Ordering::Less,
+                // If both summaries match the current Rust version or neither do, try to
+                // pick the latest version.
+                _ => s1.package_id().version().cmp(s2.package_id().version()),
+            }
+        });
+
+    match summary {
+        Some(summary) => Ok(summary.package_id()),
+        None => {
+            anyhow::bail!(
+                "could not find `{}` in registry `{}`",
+                spec,
+                source_ids.original.url()
+            )
+        }
+    }
+}
+
+fn query_summaries(
+    spec: &PackageIdSpec,
+    registry: &mut PackageRegistry<'_>,
+    source_ids: &RegistrySourceIds,
+) -> CargoResult<Vec<IndexSummary>> {
+    // Query without version requirement to get all index summaries.
+    let dep = Dependency::parse(spec.name(), None, source_ids.original)?;
+    loop {
+        // Exact to avoid returning all for path/git
+        match registry.query_vec(&dep, QueryKind::Exact) {
+            std::task::Poll::Ready(res) => {
+                break res;
+            }
+            std::task::Poll::Pending => registry.block_until_ready()?,
+        }
+    }
+}
+
+// Try to list the login and name of all owners of a crate.
+fn try_list_owners(
+    gctx: &GlobalContext,
+    source_ids: RegistrySourceIds,
+    package_name: &str,
+) -> CargoResult<Option<Vec<String>>> {
+    // Only remote registries support listing owners.
+    if !source_ids.original.is_remote_registry() {
+        return Ok(None);
+    }
+    let registry = api_registry(gctx, source_ids)?;
+    match registry {
+        Some(mut registry) => {
+            let owners = registry.list_owners(package_name)?;
+            let names = owners.iter().map(get_username).collect();
+            Ok(Some(names))
+        }
+        None => Ok(None),
+    }
+}
+
+fn get_username(u: &User) -> String {
+    format!(
+        "{}{}",
+        u.login,
+        u.name
+            .as_ref()
+            .map(|name| format!(" ({})", name))
+            .unwrap_or_default(),
+    )
+}
+
+struct RegistrySourceIds {
+    /// Use when looking up the auth token, or writing out `Cargo.lock`
+    original: SourceId,
+    /// Use when interacting with the source (querying / publishing , etc)
+    ///
+    /// The source for crates.io may be replaced by a built-in source for accessing crates.io with
+    /// the sparse protocol, or a source for the testing framework (when the replace_crates_io
+    /// function is used)
+    ///
+    /// User-defined source replacement is not applied.
+    /// Note: This will be utilized when interfacing with the registry API.
+    replacement: SourceId,
+}
+
+fn get_source_id(
```

Comment:
This function is pretty similar with https://github.com/hi-rustin/cargo/blob/476969f3cfa4ec2fd8ca202132ce0405093aa863/src/cargo/ops/registry/mod.rs#L187

I am not sure whether we should extend the original code or try to move out the `package_id` check here.

Reply:
Even if we decided to move the `package_id` check out, we still cannot use [registry](https://github.com/hi-rustin/cargo/blob/476969f3cfa4ec2fd8ca202132ce0405093aa863/src/cargo/ops/registry/mod.rs#L117) directly. Because it calls the original `get_source_id`.

Reply:
So I prefer to extend [registry](https://github.com/hi-rustin/cargo/blob/476969f3cfa4ec2fd8ca202132ce0405093aa863/src/cargo/ops/registry/mod.rs#L117) and [get_source_id](https://github.com/hi-rustin/cargo/blob/476969f3cfa4ec2fd8ca202132ce0405093aa863/src/cargo/ops/registry/mod.rs#L187) to support specify the package ID.

Reply:
@epage Do you have any thoughts about it? Thanks!

Reply:
So the idea is to extract the `get_source_id` out of `registry`?  That seems reasonable.

Reply:
I think it makes sense to move the info ops into the registry rather than extracting it out of the registry. What I really want to do is try to extend the registry API to support the use of source IDs from a specific package. I did it in:[`55db322` (#14141)](https://github.com/rust-lang/cargo/pull/14141/commits/55db322a9187aa5d099fc7b4c922948342609cdb)

I am not sure if this is a really good API design. Maybe we should consider moving the `use_package_source_id` into the RegistrySourceIds instead of exposing it directly as a tuple field?

I know I should update the comments of these APIs. I will do it after you take a quick glance to make sure I am on the right track.